docs/proposals/202005_query_logging_rate_limiting.md

```diff
@@ -0,0 +1,126 @@
+
+### Query Logging in Thanos
+
+#### Description of the Problem
+
+The Query Logging is a feature which helps in tracking all the active queries made in a database, and helps the users in knowing/debugging what queries made were resource-intensive, and how they can use the details to further optimize the database, or from the user’s perspective, to draw out the line from making such a request.
+
+A similar feature exists in Prometheus, here is a description of an active Query Log [[1]](https://www.robustperception.io/what-queries-were-running-when-prometheus-died) -
+
+```powershell
+level=info ts=2019-08-28T14:30:09.142Z caller=main.go:329 msg="Starting Prometheus" version="(version=2.12.0, branch=HEAD, revision=43acd0e2e93f9f70c49b2267efa0124f1e759e86)"
+level=info ts=2019-08-28T14:30:09.142Z caller=main.go:330 build_context="(go=go1.12.8, user=root@7a9dbdbe0cc7, date=20190818-13:53:16)"
+level=info ts=2019-08-28T14:30:09.142Z caller=main.go:331 host_details="(Linux 4.15.0-55-generic #60-Ubuntu SMP Tue Jul 2 18:22:20 UTC 2019 x86_64 mari (none))"
+level=info ts=2019-08-28T14:30:09.142Z caller=main.go:332 fd_limits="(soft=1000000, hard=1000000)"
+level=info ts=2019-08-28T14:30:09.142Z caller=main.go:333 vm_limits="(soft=unlimited, hard=unlimited)"
+level=info ts=2019-08-28T14:30:09.143Z caller=query_logger.go:74 component=activeQueryTracker msg="These queries didn't finish in prometheus' last run:" queries="[{\"query\":\"changes(changes(prometheus_http_request_duration_seconds_bucket[1h:1s])[1h:1s])\",\"timestamp_sec\":1567002604}]"
+level=info ts=2019-08-28T14:30:09.144Z caller=main.go:654 msg="Starting TSDB ..."level=info
+```
+
+Here, we can see that this log was taken before the prometheus instance died unnaturally, and the query logged under the component activeQueryTracker, logs the respective query that did the damage. Even though it does not pin-points the exact issue that created the problem, still it is quite helpful in considering this issue as one of the potential queries that led to the unnatural death of the Prometheus instance.
+
+#### Possible Approach
+
+A possible solution would be to add a Logger to track all the queries made in all of the Thanos components. Since the different components of Thanos are interconnected with the StoreAPI, it would be helpful in tracking all the queries in the global state. The diagram here represents the storeAPI and how it is connected to all the components of Thanos - 
+
+![](../img/thanos_log_limit.png)
+
+Here we can track all the active queries made using a middleware between all the requests of the StoreAPI. Here, the QueryLogger would log all the active queries, thus keeping a track of all the query requests made until now. The middleware, or rather the interceptors, would intercept all the queries and log them to a file that would contain the active logs. This logging would certainly be different from the general logging that has been mentioned, and would have the following interface - 
```

Comment:
Just make sure that even if something works for Prometheus, we need to be careful, because Thanos is distributed system., Maybe there exists simpler solution.

Reply:
Sure, I think if we can implement this local to a given component, rather than using it at the StoreAPI, this would be a useful addition. What do you think about this?